File: app/src/main/java/com/androidacademy/msk/exerciseproject/screen/news_details/NewsDetailsFragment.java
package com.androidacademy.msk.exerciseproject.screen.news_details;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.androidacademy.msk.exerciseproject.R;
import com.androidacademy.msk.exerciseproject.data.database.entity.DbNewsItem;
import com.androidacademy.msk.exerciseproject.di.Injector;
import com.androidacademy.msk.exerciseproject.utils.DateUtils;
import com.arellomobile.mvp.MvpAppCompatFragment;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.bumptech.glide.Glide;

import javax.inject.Inject;

public class NewsDetailsFragment extends MvpAppCompatFragment implements NewsDetailsView {

    private static final String KEY_ID = "KEY_ID";

    @NonNull
    private TextView titleTextView;
    @NonNull
    private ImageView imageView;
    @NonNull
    private TextView abstractTextView;
    @NonNull
    private TextView dateTextView;
    @NonNull
    private TextView timeTextView;

    @Inject
    @InjectPresenter
    public NewsDetailsPresenter presenter;

    @ProvidePresenter
    public NewsDetailsPresenter providePresenter() {
        int id = getArguments().getInt(KEY_ID);
        Injector.getInstance().getNewsItemComponent(id).inject(this);
        return presenter;
    }

    public static NewsDetailsFragment newInstance(int id) {
        Bundle args = new Bundle();
        args.putInt(KEY_ID, id);

        NewsDetailsFragment fragment = new NewsDetailsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_news_details, container, false);

        titleTextView = view.findViewById(R.id.textview_newsdetails_title);
        imageView = view.findViewById(R.id.imageview_newsdetails);
        abstractTextView = view.findViewById(R.id.textview_newsdetails_abstract);
        dateTextView = view.findViewById(R.id.textview_newsdetails_date);
        timeTextView = view.findViewById(R.id.textview_newsdetails_time);

        return view;
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem menuItem) {
        if (menuItem.getItemId() == android.R.id.home) {
            getActivity().onBackPressed();
        }
        return super.onOptionsItemSelected(menuItem);
    }

    @Override
    public void showNewsDetails(@NonNull DbNewsItem newsItem) {
        titleTextView.setText(newsItem.getTitle());
        if (newsItem.getFullsizeImageUrl() != null) {
            Glide.with(this).load(newsItem.getFullsizeImageUrl()).into(imageView);
        } else {
            imageView.setVisibility(View.GONE);
        }
        abstractTextView.setText(newsItem.getAbstractX());
        String publishedDate = newsItem.getPublishedDate();
        if (publishedDate != null) {
            String date = DateUtils.getFormattedDate(publishedDate);
            dateTextView.setText(date);
            String time = DateUtils.getFormattedTime(publishedDate, getContext());
            timeTextView.setText(time);
        }
    }
}

File: app/src/main/java/com/androidacademy/msk/exerciseproject/screen/newsdetails/NewsDetailsFragment.kt
package com.androidacademy.msk.exerciseproject.screen.newsdetails

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup

import com.androidacademy.msk.exerciseproject.R
import com.androidacademy.msk.exerciseproject.data.database.entity.DbNewsItem
import com.androidacademy.msk.exerciseproject.di.DI
import com.androidacademy.msk.exerciseproject.di.model.NewsId
import com.androidacademy.msk.exerciseproject.utils.DateUtils
import com.arellomobile.mvp.MvpAppCompatFragment
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.fragment_news_details.*

class NewsDetailsFragment : MvpAppCompatFragment(), NewsDetailsView {

    companion object {

        private const val KEY_ID = "KEY_ID"

        fun newInstance(id: Int): NewsDetailsFragment {
            val args = Bundle()
            args.putInt(KEY_ID, id)

            val fragment = NewsDetailsFragment()
            fragment.arguments = args
            return fragment
        }
    }

    @InjectPresenter
    lateinit var presenter: NewsDetailsPresenter

    @ProvidePresenter
    fun providePresenter(): NewsDetailsPresenter {

        val id = NewsId(arguments?.getInt(KEY_ID) ?: 0)
        return DI.openNewsDetailsScope(id).getInstance(NewsDetailsPresenter::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater,
                              container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        return inflater.inflate(R.layout.fragment_news_details, container, false)
    }

    override fun onDestroy() {
        super.onDestroy()
        DI.closeNewsDetailsScope()
    }

    override fun onOptionsItemSelected(menuItem: MenuItem?): Boolean {
        if (menuItem?.itemId == android.R.id.home) {
            activity?.onBackPressed()
        }
        return super.onOptionsItemSelected(menuItem)
    }


    override fun showNewsDetails(newsItem: DbNewsItem) {
        textview_newsdetails_title.text = newsItem.title
        if (newsItem.fullsizeImageUrl.isNotEmpty()) {
            Glide.with(this).load(newsItem.fullsizeImageUrl).into(imageview_newsdetails)
        } else {
            imageview_newsdetails.visibility = View.GONE
        }
        textview_newsdetails_abstract.text = newsItem.abstractX
        val publishedDate = newsItem.publishedDate
        if (publishedDate.isNotEmpty()) {
            val date = DateUtils.getFormattedDate(publishedDate)
            textview_newsdetails_date.text = date
            context?.let {
                val time = DateUtils.getFormattedTime(publishedDate, it)
                textview_newsdetails_time.text = time
            }
        }
    }
}
